amazon/ionhash/hasher.py

```diff
@@ -0,0 +1,462 @@
+    support a hash function for use by `hash_reader` or `hash_writer`."""
+    @abstractmethod
+    def update(self, _bytes):
+        """Updates the hash function with the specified _bytes."""
+        pass
+
+    @abstractmethod
+    def digest(self):
+        """Returns a digest of the accumulated bytes passed to `update`, and resets the `IonHasher`
+        to its initial state."""
+        pass
+
+
+class _HashlibHash(IonHasher):
+    """Implements the expected hash function methods for the specified algorithm using hashlib."""
+    def __init__(self, algorithm):
+        self._algorithm = algorithm
+        self._hasher = hashlib.new(self._algorithm)
+
+    def update(self, _bytes):
+        self._hasher.update(_bytes)
+
+    def digest(self):
+        digest = self._hasher.digest()
+        self._hasher = hashlib.new(self._algorithm)
+        return digest
+
+
+def hash_reader(reader, hash_function_provider):
+    """Provides a coroutine that wraps an ion-python reader and adds Ion Hash functionality.
+
+    The given coroutine yields `bytes` when given ``HashEvent.DIGEST``.  Otherwise, the
+    couroutine's behavior matches that of the wrapped reader.
+
+    Notes:
+        The coroutine translates any `SKIP_EVENT`s into a series of `NEXT_EVENT`s in order
+        to ensure that the hash correctly includes any subsequent or nested values.
+
+    Args:
+        reader(couroutine):  An ion-python reader coroutine.
+        hash_function_provider(function):  A function that returns a function that produces
+            ``IonHasher`` instances when called.  Note that multiple ``IonHasher`` instances
+            may be required to hash a single value (depending on the type of the Ion value).
+
+    Yields:
+        bytes:  The result of hashing.
+        other values:  As defined by the provided reader coroutine.
+    """
+    hr = _hasher(_hash_reader_handler, reader, hash_function_provider)
+    next(hr)    # prime the coroutine
+    return hr
+
+
+def hash_writer(writer, hash_function_provider):
+    """Provides a coroutine that wraps an ion-python writer and adds Ion Hash functioality.
+
+    The given coroutine yields `bytes` when given ``HashEvent.DIGEST``.  Otherwise, the
+    couroutine's behavior matches that of the wrapped writer.
+
+    Args:
+        writer(coroutine):  An ion-python writer coroutine.
+        hash_function_provider(function):  A function that returns a function that produces
+            ``IonHasher`` instances when called.  Note that multiple ``IonHasher`` instances
+            may be required to hash a single value (depending on the type of the Ion value).
+
+    Yields:
+        bytes:  The result of hashing.
+        other values:  As defined by the provided writer coroutine.
+    """
+    hw = _hasher(_hash_writer_handler, writer, hash_function_provider)
+    next(hw)    # prime the coroutine
+    return hw
+
+
+def _hasher(handler, delegate, hash_function_provider):
```

Comment:
In ion-python we use the [`@coroutine`](https://github.com/amzn/ion-python/blob/master/amazon/ion/util.py#L166) decorator to avoid having to "prime" the coroutine with `next` (e.g. line [91](https://github.com/amzn/ion-python/blob/master/amazon/ion/util.py#L166)). Consider using here too. I also think it's kind of nice because it's easy to see which functions are intended to be coroutines.